### ISSUE TITLE
Feature Request: Support for top-level await

### ISSUE THREAD

**Issue by almercier** [jshint/jshint#3447]
The [proposal is in stage 3](https://tc39.es/proposal-top-level-await/) now, and in [the V8 engine](https://v8.dev/features/top-level-await). it's supported by [Chrome through a dev flag](https://bugs.chromium.org/p/v8/issues/detail?id=9344) and [babel also has a plugin for it ](https://www.npmjs.com/package/@babel/plugin-syntax-top-level-await)

It would be really nice to have expose an option to allow this in jshint.

**Comment by jugglinmike** [comment]
Thanks for the report! Although the feature's implementation in other projects doesn't influence our priorities, [it's position in TC-39's standardization process certainly does](https://jshint.com/blog/new-lang-features/). This is well past due!

**Comment by Dreamsavior** [comment]
@jugglinmike 
Is there any chance that this issue will be fixed soon?
Or, do you have any idea or hint on how to fix this? Perhaps there is a work around may be?
![image](https://user-images.githubusercontent.com/1912407/147482250-1adfe0b9-6132-4ff0-ba76-a7b738db2196.png)

**Comment by pr4xx** [comment]
+1

**Comment by almercier** [comment]
@dreamsavior 
Our workaround currently is wrapping the logic in an async anonymous function and immediately executing it.

```Javascript
(async () {
   // ...
    var translated = await translator.translate([textBody]);
    //...
})();

**Comment by ehoogeveen-medweb** [comment]
One issue with that workaround when working with ES modules in Node.js: An asynchronous function waiting on promises isn't enough to keep the process alive unless you use top-level await to await that function (see https://github.com/nodejs/node/issues/22088). So you end up needing another workaround like https://github.com/nodejs/node/issues/22088#issuecomment-895440267 to keep the process from silently exiting early.

**Comment by Othmane-ElAlami** [comment]
2 years later.. still no way to ignore this "error" ?

**Comment by danbopes** [comment]
This is incredibly painful. There really should be a way to allow await in the top-level. This is a blocker for me ATM, and is causing me to just disable jshint all together.

**Comment by JYone3A** [comment]
There's a bug actually, this isn't just a feature request/proposal.

When setting `module` to true, top level `await` should not throw an error, as that's how it actually works with ESM.

**Comment by JYone3A** [comment]
Hmmm according to this line it should work, if I understand that correctly ...? Nor sure what I'm doing wrong.

https://github.com/jshint/jshint/blob/main/src/jshint.js#L177

Anyway, wouldn't it be simple to extend this line to something like:

```javascript
 if (token.id === "await" && (!(context & prodParams.async) && !state.option.module && !state.option.asyncContext)) {
```

No idea about the code so maybe its not that easy ;)

**Comment by JYone3A** [comment]
Played around a little, first I changed [line 6655](https://github.com/jshint/jshint/blob/main/src/jshint.js#L6655)

```javascript
// Previous:
statements(0);

// Now:
statements(state.option.module ? prodParams.async : 0);
```

And then I changed [line 5434](https://github.com/jshint/jshint/blob/main/src/jshint.js#L5434):
```javascript
// Previous:
if (!state.funct["(params)"]) {

// Now:
if (!state.funct["(params)"] && !state.funct["(global)"]) {
```

Now no more error pops up when using `await` on the top level, but still errors pop up when defining a non-async function in there where `await` is used.

Does that go in the right direction?

**Comment by dreamsavior** [comment]
Thank you @JYone3A.

With the same logic inspired by @JYone3A solution, I successfully add an option to jshint to enable top level `await`.

Firstly introduce the new option to the `options.js`  [line 788](https://github.com/jshint/jshint/blob/0a5644f8f529e252e7dd0c0d54334ae435b13de0/src/options.js#L788C23-L788C23)
```JS
    phantom     : true,

    /**
     * Enable top level await
     */
    topLevelAwait : true
```

And then change `jshint.js` at [line 6655]([line 6655](https://github.com/jshint/jshint/blob/main/src/jshint.js#L6655))
```JS
// Previous:
statements(0);

// Now:
statements(state.option.topLevelAwait ? prodParams.async : 0);
```

And [line 5434](https://github.com/jshint/jshint/blob/main/src/jshint.js#L5434):
```JS
// Previous:
if (!state.funct["(params)"]) {

// Now:
if (!state.funct["(params)"] && !state.option.topLevelAwait) {
```

I also successfully apply the fix on Ace editor:
<img width="945" alt="image" src="https://github.com/jshint/jshint/assets/1912407/c1a2102e-cf43-4363-a639-4dd6e584681f">

**Comment by jugglinmike** [comment]
Thanks for working on this, you two! Since this now a standardized behavior, it has well-defined semantics which we can use to enable it--no new linting options required. Top-level `await` should be automatically enabled when users have set the `esVersion` option to `13` (or `"2022"`) and the `module` linting option to `true`.

**Comment by dreamsavior** [comment]
@jugglinmike 
Thank you, that's great. However, in fact, we needed an option to explicitly enable top-level await even prior to ES 2022. For example, for Vue 3 code snippets or just to make it work for ACE editor, for whatever purpose people see fit. This case exists in a real-life application.

**Comment by JYone3A** [comment]
> ```js
> // Previous:
> statements(0);
> 
> // Now:
> statements(state.option.topLevelAwait ? prodParams.async : 0);
> ```

This should probably be `statements(state.option.topLevelAwait || state.option.module ? prodParams.async : 0);`

> ```js
> // Previous:
> if (!state.funct["(params)"]) {
> 
> // Now:
> if (!state.funct["(params)"] && !state.option.topLevelAwait) {
> ```

I do believe my updated version above is better, i.e. `if (!state.funct["(params)"] && !state.funct["(global)"]) {` as it would only match top level scopes, at least if I did correctly understand what `(global)` means. Not sure.

@jugglinmike while for me personally it is sufficient if the `module` flag enables top level await, I agree with @dreamsavior that there's a need for a separate top level await flag (~~e.g. for `.mjs` scrips run with nodejs~~ ok no that was nonsense).

**Comment by JYone3A** [comment]
The reason I think a separate top level await is needed is that there are two differences to a module:
1. not necessarily strict mode
2. no `export`

I guess jshint could or should enforce at least one export statement in a module, right? So using the module logic would not work in @dreamsavior's case

**Comment by JYone3A** [comment]
> Top-level `await` should be automatically enabled when users have set the `esVersion` option to `13` (or `"2022"`) and the `module` linting option to `true`.

Do you know when this will land? Would a pull request help, and if yes, was the solution above somewhat in the right direction?

**Comment by coolaj86** [comment]
@jugglinmike What amount of money would it take to buy your time to get this implemented?

**Comment by jaapmurre** [comment]
I just ran into this problem as well.

**Comment by jaapmurre** [comment]
I created a work-around so the toplevel await error is ignored. A limitation is that other mistakes in that line are no longer signaled. This is my CodeMirror code at the moment.

```javascript
// Codemirror EditorView instance ref
const view = shallowRef(EditorView)
const handleReady = (payload:any) => {
  view.value = payload.view
}

import { linter } from '@codemirror/lint';
import { lintGutter } from '@codemirror/lint';

const lintOptions = {
    esversion: 13,       // Set ECMAScript version to 13 (or "2022")
    browser: true,       // Asume browser environment vs NodeJS (node: true)
    asi: true,           // Less strict
    expr: true,          // Less strict
    undef: false,        // Less strict    
    //    topLevelAwait : true, // Allow top-level await, not implemented in JSHint
  };

import { JSHINT as jshint } from 'jshint'; 
(<any>window).JSHINT = jshint; 

const jsLinter = (lintOptions:any) => {
  return linter((view) => {
    const diagnostics = <any>[];
    const codeText = view.state.doc.toJSON();

    // For now, I am only generating errors for user code, not including header/footer
    // jshint( [].concat(props.header.split('\n'),props.header_marker.split('\n'), 
    //        codeText, props.footer_marker.split('\n'), props.footer.split('\n')) , lintOptions);
    jshint(codeText,lintOptions);
    const errors = jshint?.data().errors;

    if (errors) {
      console.log("ERRORS: ",errors);
    }

    if (errors && errors.length > 0) {
      errors.forEach((error:any) => {
        // For now, I am only generating errors for user code, not including header/footer
        // const selectedLine = view.state.doc.line(error.line 
        //           - (props.header + props.header_marker).split('\n').length - 1);
        const selectedLine = view.state.doc.line(error.line);

        if (selectedLine.number <= codeText.length) {

          if (selectedLine.text.trim().indexOf('await') === 0) {
            // If the line starts with 'await', we don't want to show an error
            // because it is likely a top-level await in a module
            return;
          }

          console.log("Error line: ",selectedLine.number);
          const diagnostic = {
            from: selectedLine.from,
            to: selectedLine.to,
            severity: 'error',
            message: error.reason,
          };

          diagnostics.push(diagnostic);
        }
      });
    }
    return diagnostics;
  });
};

let extensions: Extension[];

function updateMode(newMode:string) {
  if (newMode === 'Python') {
    extensions =  [basicSetup, python(), lintGutter()];
  } else if (newMode === 'JavaScript') {
    extensions = [basicSetup, javascript(), jsLinter(lintOptions), lintGutter()];
  }
}

updateMode(props.mode);
watch(() => props.mode, (newMode) => { updateMode(newMode); }); 

```
